### ISSUE TITLE
Applications of Graph Matching

### ISSUE THREAD

**Issue by dlee138** [openconnectome/Statistical-Connectomics-Class#214]
So what exactly is graph matching useful for in the real world?

**Comment by SandyaS72** [comment]
I would be interested in this as well for fields outside of research/science?

**Comment by imichaelnorris** [comment]
In the biochem world, we used graph matching to assign experimental and predicted molecules for an RNA in some optimization algorithm.  We ran the graph matching, calculated a score, and permuted the graphs (genetic algorithm) to try to converge to an optimal assignment.

**Comment by imichaelnorris** [comment]
I'm sure NSA uses graph matching to detect terrorist cells.  Say that network traffic is a graph.  Maybe you are looking for traffic starting from 1 node (a person) that ends up at a propaganda website or communicates with another suspicious person.  If you have a collection of these "high value targets" (nodes), then you can do a graph match on a person with edges to their internet traffic. If there's a person with a lot of edges to the high value targets, then the graph matches.

One example I was reading.  Say that there's one person who bought 500 pounds of fertilizer, and another person who rented a truck.  Alone, it's probably not a big deal.  But if you learn that they were roommates -- call this an edge between them -- then you should pay more attention.  If you're labeling relationships and activity, and you can produce a graph that contains multiple people possessing everything they need to commit a terrorist attack, you can run a graph match on your known data.  If you come up with any hits, you should investigate.

**Comment by whock** [comment]
I think in general graph matching is the tool you use to translate one graph into another when you have reason to believe a priori that they have some common ground truth, e.g. what @imichaelnorris said about the terrorist cells. Consider the graphs of two different social networking sites, say twitter and snapchat. These are separate graphs in the sense that if you were to visualize them they'd have different nodes, edges, etc. But they share common ground truth - many of the same people use both and are connected to the same people in both. So if I know some nodes and edges in twitter correspond to the members of an ISIS cell and their connections, how can I use this to discover the same cell on snapchat (provided you've hacked it to find numbers, i.e. provided you're the NSA). The key insight is that the people connected on one network are probably connected on the other so you use knowledge of one to infer structure of the other. You're looking at the cell from two different perspectives and graph matching is a way to reconcile / translate their perspectives.

Graph matching has many more applications, some of which are listed here:
![graphmatching](https://cloud.githubusercontent.com/assets/10751093/7220454/7ec60364-e695-11e4-8b14-cabc1b46c3ec.png)

(taken from http://www.cs.odu.edu/~mhalappa/matching/applications.html which has in depth bibliography)

**Comment by jtmatterer** [comment]
In statistics on graphs, the term "graph matching" refers to what I presented last week, however, in combinatorial optimization, there is a completely different problem called "graph matching", which is essentially finding a collection of edges in a single graph so that no two vertices are adjacent and maximizing some objective function. Unfortunately the applications @whock mentioned are for the latter problem.

Adding on to what @imichaelnorris wrote, a partially outdated discussion of applications is given in section 3 of http://www.worldscientific.com/doi/abs/10.1142/S0218001404003228.

**Comment by yaxigeigei** [comment]
I can think of an application on image registration. Image registration has very broad applications including in neuroscience (e.g. for two-photon volume imaging in moving animals). 
Features of an image can be computed using fairly straightforward computer vision algorithm such as those provided in MATLAB toolbox. There are existing algorithms that can take constant features in different image as reference points for fast and accurate registration. However, as the image metamorphoses, features will change especially when the signal is noisy. We can make identified features and their spatial relationship to the image into graphs and matching them will unify those reference points for further registration.
To my knowledge, currently there is no good way to registrate thick 4D(3D+time) two-photon images.

**Comment by jovo** [comment]
de-anonymization sometimes.
look it up, pretty cool :)

On Tue, Apr 21, 2015 at 11:41 PM, Duo Xu <notifications@github.com> wrote:

>  I can think of an application on image registration. Image registration
> has very broad applications including in neuroscience (e.g. for two-photon
> volume imaging in moving animals).
> Features of an image can be computed using fairly straightforward computer
> vision algorithm such as those provided in MATLAB toolbox. There are
> existing algorithms that can take constant features in different image as
> reference points for fast and accurate registration. However, as the image
> metamorphoses, features will change especially when the signal is noisy. We
> can make identified features and their spatial relationship to the image
> into graphs and matching them will unify those reference points for further
> registration.
> To my knowledge, currently there is no good way to registrate thick
> 4D(3D+time) two-photon images.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openconnectome/Statistical-Connectomics-Class/issues/214#issuecomment-95017229>
> .
>



-- 
the glass is all full:  half water, half air.
openconnecto.me, jovo.me, my calendar
<https://www.google.com/calendar/embed?src=joshuav%40gmail.com&ctz=America/New_York>

**Comment by whock** [comment]
I have brought shame upon my ancestors :(

But matching, e.g. , social networks to find a structure in one given known structure in another is a good example of the graph matching @jtmatterer presented, right? And is this the de-anonymization Jovo mentioned?

**Comment by mblohr** [comment]
Yes.  Here are some examples of de-anonymizing social networks by graph matching:  
https://www.cs.utexas.edu/~shmat/shmat_oak09.pdf
http://arxiv.org/pdf/1411.7296v1.pdf
